# Problem Statement:

Many social programs have a hard time making sure the right people are given enough aid. It’s tricky when a program focuses on the poorest segment of the population. This segment of population can’t provide the necessary income and expense records to prove that they qualify.

In Latin America, a popular method called Proxy Means Test (PMT) uses an algorithm to verify income qualification. With PMT, agencies use a model that considers a family’s observable household attributes like the material of their walls and ceiling or the assets found in their homes to classify them and predict their level of need. While this is an improvement, accuracy remains a problem as the region’s population grows and poverty declines.

The Inter-American Development Bank (IDB) believes that new methods beyond traditional econometrics, based on a dataset of Costa Rican household characteristics, might help improve PMT’s performance.

In [1]:
import os
import warnings
warnings.filterwarnings('ignore')

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [3]:
train = pd.read_csv('D:/DS R/train.csv')
test = pd.read_csv('D:/DS R/test.csv')

# Explore the dataset before moving further

In [4]:
print('Shape of train dataset is {}'.format(train.shape))
print('Shape of test dataset is {}'.format(test.shape))

Shape of train dataset is (9557, 143)
Shape of test dataset is (23856, 142)


# Identify the target variable

In [5]:
for i in train.columns:
    if i not in test.columns:
        print("Our Target variable is {}".format(i))

Our Target variable is Target


# Understand the type of data

In [6]:
print(train.dtypes.value_counts())

int64      130
float64      8
object       5
dtype: int64


In [7]:
print(train.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9557 entries, 0 to 9556
Columns: 143 entries, Id to Target
dtypes: float64(8), int64(130), object(5)
memory usage: 10.4+ MB
None


# Explore each different types of datasets

In [8]:
for i in train.columns:
    a=train[i].dtype
    if a == 'object':
        print(i)

Id
idhogar
dependency
edjefe
edjefa


### Below is Data dictionary for above object variables

ID = Unique ID

idhogar, Household level identifier

dependency, Dependency rate, calculated = (number of members of the household younger than 19 or older than 64)/(number of member of household between 19 and 64)

edjefe, years of education of male head of household, based on the interaction of escolari (years of education), head of household and gender, yes=1 and no=0

edjefa, years of education of female head of household, based on the interaction of escolari (years of education), head of household and gender, yes=1 and no=0

# Drop Id variable

In [9]:
train.drop(['Id','idhogar'],axis=1,inplace=True)

In [10]:
train['dependency'].value_counts()

yes          2192
no           1747
.5           1497
2             730
1.5           713
.33333334     598
.66666669     487
8             378
.25           260
3             236
4             100
.75            98
.2             90
1.3333334      84
.40000001      84
2.5            77
5              24
.80000001      18
3.5            18
1.25           18
2.25           13
.71428573      12
1.75           11
.22222222      11
.83333331      11
1.2            11
.2857143        9
.60000002       8
1.6666666       8
6               7
.16666667       7
Name: dependency, dtype: int64

# Convert the object variables into numerical data

In [11]:
def map(i):
    
    if i=='yes':
        return(float(1))
    elif i=='no':
        return(float(0))
    else:
        return(float(i))

In [12]:
train['dependency']=train['dependency'].apply(map)

In [13]:
for i in train.columns:
    a=train[i].dtype
    if a == 'object':
        print(i)

edjefe
edjefa


In [14]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9557 entries, 0 to 9556
Columns: 141 entries, v2a1 to Target
dtypes: float64(9), int64(130), object(2)
memory usage: 10.3+ MB


   Now the data is in numerical form

# Identify variable with 0 varinace

In [15]:
var_df=pd.DataFrame(np.var(train,0),columns=['variance'])
var_df.sort_values(by='variance').head(15)
print('Below are columns with variance 0.')
col=list((var_df[var_df['variance']==0]).index)
print(col)

Below are columns with variance 0.
['elimbasu5']


elimbasu5 : 1 if rubbish disposal mainly by throwing in river, creek or sea.

Interpretation :From above it is shown that all values of elimbasu5 is same so there is no variablity in dataset therefor we will drop this variable

# Check if there are any biases in your dataset

In [16]:
contingency_tab=pd.crosstab(train['r4t3'],train['hogar_total'])
Observed_Values=contingency_tab.values
import scipy.stats
b=scipy.stats.chi2_contingency(contingency_tab)
Expected_Values = b[3]
no_of_rows=len(contingency_tab.iloc[0:2,0])
no_of_columns=len(contingency_tab.iloc[0,0:2])
df=(no_of_rows-1)*(no_of_columns-1)
print("Degree of Freedom:-",df)
from scipy.stats import chi2
chi_square=sum([(o-e)**2./e for o,e in zip(Observed_Values,Expected_Values)])
chi_square_statistic=chi_square[0]+chi_square[1]
print("chi-square statistic:-",chi_square_statistic)
alpha=0.05
critical_value=chi2.ppf(q=1-alpha,df=df)
print('critical_value:',critical_value)
p_value=1-chi2.cdf(x=chi_square_statistic,df=df)
print('p-value:',p_value)
print('Significance level: ',alpha)
print('Degree of Freedom: ',df)
print('chi-square statistic:',chi_square_statistic)
print('critical_value:',critical_value)
print('p-value:',p_value)
if chi_square_statistic>=critical_value:
    print("Reject H0,There is a relationship between 2 categorical variables")
else:
    print("Retain H0,There is no relationship between 2 categorical variables")
    
if p_value<=alpha:
    print("Reject H0,There is a relationship between 2 categorical variables")
else:
    print("Retain H0,There is no relationship between 2 categorical variables")

Degree of Freedom:- 1
chi-square statistic:- 17022.072400560897
critical_value: 3.841458820694124
p-value: 0.0
Significance level:  0.05
Degree of Freedom:  1
chi-square statistic: 17022.072400560897
critical_value: 3.841458820694124
p-value: 0.0
Reject H0,There is a relationship between 2 categorical variables
Reject H0,There is a relationship between 2 categorical variables


Therefore,variables ('r4t3','hogar_total') have relationship between them. For good result we can use any one of them.

In [17]:
contingency_tab=pd.crosstab(train['tipovivi3'],train['v2a1'])
Observed_Values=contingency_tab.values
import scipy.stats
b=scipy.stats.chi2_contingency(contingency_tab)
Expected_Values = b[3]
no_of_rows=len(contingency_tab.iloc[0:2,0])
no_of_columns=len(contingency_tab.iloc[0,0:2])
df=(no_of_rows-1)*(no_of_columns-1)
print("Degree of Freedom:-",df)
from scipy.stats import chi2
chi_square=sum([(o-e)**2./e for o,e in zip(Observed_Values,Expected_Values)])
chi_square_statistic=chi_square[0]+chi_square[1]
print("chi-square statistic:-",chi_square_statistic)
alpha=0.05
critical_value=chi2.ppf(q=1-alpha,df=df)
print('critical_value:',critical_value)
p_value=1-chi2.cdf(x=chi_square_statistic,df=df)
print('p-value:',p_value)
print('Significance level: ',alpha)
print('Degree of Freedom: ',df)
print('chi-square statistic:',chi_square_statistic)
print('critical_value:',critical_value)
print('p-value:',p_value)
if chi_square_statistic>=critical_value:
    print("Reject H0,There is a relationship between 2 categorical variables")
else:
    print("Retain H0,There is no relationship between 2 categorical variables")
    
if p_value<=alpha:
    print("Reject H0,There is a relationship between 2 categorical variables")
else:
    print("Retain H0,There is no relationship between 2 categorical variables")

Degree of Freedom:- 1
chi-square statistic:- 54.04781105990782
critical_value: 3.841458820694124
p-value: 1.9562129693895258e-13
Significance level:  0.05
Degree of Freedom:  1
chi-square statistic: 54.04781105990782
critical_value: 3.841458820694124
p-value: 1.9562129693895258e-13
Reject H0,There is a relationship between 2 categorical variables
Reject H0,There is a relationship between 2 categorical variables


Therefore,variables ('tipovivi3','v2a1') have relationship between them. For good result we can use any one of them.

In [18]:
contingency_tab=pd.crosstab(train['v18q'],train['v18q1'])
Observed_Values=contingency_tab.values
import scipy.stats
b=scipy.stats.chi2_contingency(contingency_tab)
Expected_Values = b[3]
no_of_rows=len(contingency_tab.iloc[0:2,0])
no_of_columns=len(contingency_tab.iloc[0,0:2])
df=(no_of_rows-1)*(no_of_columns-1)
print("Degree of Freedom:-",df)
from scipy.stats import chi2
chi_square=sum([(o-e)**2./e for o,e in zip(Observed_Values,Expected_Values)])
chi_square_statistic=chi_square[0]+chi_square[1]
print("chi-square statistic:-",chi_square_statistic)
alpha=0.05
critical_value=chi2.ppf(q=1-alpha,df=df)
print('critical_value:',critical_value)
p_value=1-chi2.cdf(x=chi_square_statistic,df=df)
print('p-value:',p_value)
print('Significance level: ',alpha)
print('Degree of Freedom: ',df)
print('chi-square statistic:',chi_square_statistic)
print('critical_value:',critical_value)
print('p-value:',p_value)
if chi_square_statistic>=critical_value:
    print("Reject H0,There is a relationship between 2 categorical variables")
else:
    print("Retain H0,There is no relationship between 2 categorical variables")
    
if p_value<=alpha:
    print("Reject H0,There is a relationship between 2 categorical variables")
else:
    print("Retain H0,There is no relationship between 2 categorical variables")

Degree of Freedom:- 0
chi-square statistic:- 0.0
critical_value: nan
p-value: nan
Significance level:  0.05
Degree of Freedom:  0
chi-square statistic: 0.0
critical_value: nan
p-value: nan
Retain H0,There is no relationship between 2 categorical variables
Retain H0,There is no relationship between 2 categorical variables


Therefore,variables ('v18q','v18q1') have relationship between them. For good result we can use any one of them.

Conclusion : Therefore, there is bias in our dataset.

In [19]:
train.drop('r4t3',axis=1,inplace=True)

# Check if there is a house without a family head

"parentesco1" =1 if household head

In [20]:
train.parentesco1.value_counts()

0    6584
1    2973
Name: parentesco1, dtype: int64

In [21]:
pd.crosstab(train['edjefa'],train['edjefe'])

edjefe   10   11   12   13   14   15   16   17  18  19  ...  21    3    4  \
edjefa                                                  ...                 
10        0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
11        0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
12        0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
13        0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
14        0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
15        0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
16        0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
17        0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
18        0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
19        0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
2         0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
20        0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
21        0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
3         0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
4         0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
5         0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
6         0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
7         0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
8         0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
9         0    0    0    0    0    0    0    0   0   0  ...   0    0    0   
no      111  751  113  103  208  285  134  202  19  14  ...  43  307  137   
yes       0    0    0    0    0    0    0    0   0   0  ...   0    0    0   

edjefe    5     6    7    8    9   no  yes  
edjefa                                      
10        0     0    0    0    0   96    0  
11        0     0    0    0    0  399    0  
12        0     0    0    0    0   72    0  
13        0     0    0    0    0   52    0  
14        0     0    0    0    0  120    0  
15        0     0    0    0    0  188    0  
16        0     0    0    0    0  113    0  
17        0     0    0    0    0   76    0  
18        0     0    0    0    0    3    0  
19        0     0    0    0    0    4    0  
2         0     0    0    0    0   84    0  
20        0     0    0    0    0    2    0  
21        0     0    0    0    0    5    0  
3         0     0    0    0    0  152    0  
4         0     0    0    0    0  136    0  
5         0     0    0    0    0  176    0  
6         0     0    0    0    0  947    0  
7         0     0    0    0    0  179    0  
8         0     0    0    0    0  217    0  
9         0     0    0    0    0  237    0  
no      222  1845  234  257  486  435  123  
yes       0     0    0    0    0   69    0  

[22 rows x 22 columns]

Interpretation : Above cross tab shows 0 male head and 0 female head which implies that there are 435 families with no family head.

# Count how many null values are existing in columns

In [22]:
train.isna().sum().value_counts()

0       135
5         2
7928      1
6860      1
7342      1
dtype: int64

## Identify number of null values in Target variable

In [23]:
train['Target'].isna().sum()

0

Interpretation : There are no null values in Target variable. Now lets proceed further and identify and fillna of other variable.

In [24]:
float_col=[]
for i in train.columns:
    a=train[i].dtype
    if a == 'float64':
        float_col.append(i)
print(float_col)

['v2a1', 'v18q1', 'rez_esc', 'dependency', 'meaneduc', 'overcrowding', 'SQBovercrowding', 'SQBdependency', 'SQBmeaned']


In [25]:
train[float_col].isna().sum()

v2a1               6860
v18q1              7342
rez_esc            7928
dependency            0
meaneduc              5
overcrowding          0
SQBovercrowding       0
SQBdependency         0
SQBmeaned             5
dtype: int64

In [26]:
train['v18q1'].value_counts()

1.0    1586
2.0     444
3.0     129
4.0      37
5.0      13
6.0       6
Name: v18q1, dtype: int64

In [27]:
pd.crosstab(train['tipovivi1'],train['v2a1'])

v2a1       0.0        12000.0    13000.0    14000.0    15000.0    16000.0    \
tipovivi1                                                                     
0                 29          3          4          3          3          2   

v2a1       17000.0    20000.0    23000.0    25000.0    ...  570540.0   \
tipovivi1                                              ...              
0                  4         22          5         21  ...         25   

v2a1       600000.0   620000.0   684648.0   700000.0   770229.0   800000.0   \
tipovivi1                                                                     
0                 11          3          3          7          3          4   

v2a1       855810.0   1000000.0  2353477.0  
tipovivi1                                   
0                 11          7          2  

[1 rows x 157 columns]

In [28]:
pd.crosstab(train['v18q1'],train['v18q'])

v18q      1
v18q1      
1.0    1586
2.0     444
3.0     129
4.0      37
5.0      13
6.0       6

Interpretation and action : 'v2a1', 'v18q1', 'rez_esc' have more than 50% null values, because for v18q1, there are families with their own house so they won't pay rent in that case it should be 0 and similar is for v18q1 there can be families with 0 tablets.

Istead we can drop a column tipovivi3,v18q

tipovivi3, =1 rented
v18q, owns a tablet
as v2a1 alone can show both **as v18q1 alone can show that if respondent owns a tablet or not

In [29]:
train['v2a1'].fillna(0,inplace=True)
train['v18q1'].fillna(0,inplace=True)

In [30]:
train.drop(['tipovivi3', 'v18q','rez_esc','elimbasu5'],axis=1,inplace=True)

In [31]:
train['meaneduc'].fillna(np.mean(train['meaneduc']),inplace=True)
train['SQBmeaned'].fillna(np.mean(train['SQBmeaned']),inplace=True)
print(train.isna().sum().value_counts())

0    136
dtype: int64


In [32]:
int_col=[]
for i in train.columns:
    a=train[i].dtype
    if a == 'int64':
        int_col.append(i)
print(int_col)

['hacdor', 'rooms', 'hacapo', 'v14a', 'refrig', 'r4h1', 'r4h2', 'r4h3', 'r4m1', 'r4m2', 'r4m3', 'r4t1', 'r4t2', 'tamhog', 'tamviv', 'escolari', 'hhsize', 'paredblolad', 'paredzocalo', 'paredpreb', 'pareddes', 'paredmad', 'paredzinc', 'paredfibras', 'paredother', 'pisomoscer', 'pisocemento', 'pisoother', 'pisonatur', 'pisonotiene', 'pisomadera', 'techozinc', 'techoentrepiso', 'techocane', 'techootro', 'cielorazo', 'abastaguadentro', 'abastaguafuera', 'abastaguano', 'public', 'planpri', 'noelec', 'coopele', 'sanitario1', 'sanitario2', 'sanitario3', 'sanitario5', 'sanitario6', 'energcocinar1', 'energcocinar2', 'energcocinar3', 'energcocinar4', 'elimbasu1', 'elimbasu2', 'elimbasu3', 'elimbasu4', 'elimbasu6', 'epared1', 'epared2', 'epared3', 'etecho1', 'etecho2', 'etecho3', 'eviv1', 'eviv2', 'eviv3', 'dis', 'male', 'female', 'estadocivil1', 'estadocivil2', 'estadocivil3', 'estadocivil4', 'estadocivil5', 'estadocivil6', 'estadocivil7', 'parentesco1', 'parentesco2', 'parentesco3', 'parentesco

In [33]:
train[int_col].isna().sum().value_counts()

0    126
dtype: int64

Interpretation : Now there is no null value in our datset.

In [34]:
train.Target.value_counts()

4    5996
2    1597
3    1209
1     755
Name: Target, dtype: int64

## Set the poverty level of the members and the head of the house same in a family.

Now for people below poverty level can be people paying less rent and don't own a house. and it also depends on whether a house is in urban area or rural area.

In [35]:
Poverty_level=train[train['v2a1'] !=0]

In [36]:
Poverty_level.shape

(2668, 136)

In [37]:
poverty_level=Poverty_level.groupby('area1')['v2a1'].apply(np.median)

In [38]:
poverty_level

area1
0     80000.0
1    140000.0
Name: v2a1, dtype: float64

For rural area level if people paying rent less than 8000 is under poverty level.

For Urban area level if people paying rent less than 140000 is under poverty level.

In [39]:
def povert(x):
    if x<8000:
        return('Below poverty level')
    
    elif x>140000:
        return('Above poverty level')
    elif x<140000:
        return('Below poverty level: Ur-ban ; Above poverty level : Rural ')

In [40]:
c=Poverty_level['v2a1'].apply(povert)

In [41]:
c.shape

(2668,)

In [42]:
pd.crosstab(c,Poverty_level['area1'])

area1                                                 0     1
v2a1                                                         
Above poverty level                                 139  1103
Below poverty level: Ur-ban ; Above poverty lev...  306  1081

Interpretation :

There are total 1242 people above poverty level independent of area whether rural or Urban
Remaining 1111 people level depends on their area
Rural :

Above poverty level= 445

Urban :

Above poverty level =1103

Below poverty level=1081

In [43]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

In [44]:
X_data=train.drop('Target',axis=1)
Y_data=train.Target

In [45]:
X_data_col=X_data.columns

In [46]:
from sklearn.preprocessing import StandardScaler
SS=StandardScaler()
X_data_1=SS.fit_transform(X_data)
X_data_1=pd.DataFrame(X_data_1,columns=X_data_col)

ValueError: could not convert string to float: 'no'